profiling/src/internal/profile/interning_api/mod.rs

```diff
@@ -0,0 +1,234 @@
+            self.sample_types.len(),
+            values.len(),
+        );
+        let stacktrace = stacktrace.get(self.generation)?;
+        let labels = labels.get(self.generation)?;
+
+        self.observations
+            .add(Sample::new(labels, stacktrace), timestamp, values)
+    }
+
+    pub fn intern_stacktrace(
+        &mut self,
+        locations: &[GenerationalId<LocationId>],
+    ) -> anyhow::Result<GenerationalId<StackTraceId>> {
+        let locations = locations
+            .iter()
+            .map(|l| l.get(self.generation))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+        let stacktrace = StackTrace { locations };
+        let id = self.stack_traces.dedup(stacktrace);
+        Ok(GenerationalId::new(id, self.generation))
+    }
+
+    pub const INTERNED_EMPTY_STRING: GenerationalId<StringId> =
+        GenerationalId::new_immortal(StringId::ZERO);
+
+    pub fn intern_string(&mut self, s: &str) -> anyhow::Result<GenerationalId<StringId>> {
+        if s.is_empty() {
+            Ok(Self::INTERNED_EMPTY_STRING)
+        } else {
+            Ok(GenerationalId::new(self.intern(s), self.generation))
+        }
+    }
+
+    pub fn intern_strings(
+        &mut self,
+        s: &[&str],
+        out: &mut [GenerationalId<StringId>],
+    ) -> anyhow::Result<()> {
+        anyhow::ensure!(s.len() == out.len());
+        for i in 0..s.len() {
+            out[i] = self.intern_string(s[i])?;
+        }
+        Ok(())
+    }
+
+    // Simple synchronization between samples and profile rotation/export.
+    // Interning a sample may require several calls to the profiler to intern intermediate values,
+    // which are not inherently atomic.  Since these intermediate values are tied to a particular
+    // profiler generation, and are invalidated when the generation changes, some coordination must
+    // occur between sampling and profile rotation/export.
+    // When the generation changes, one of three things can happen:
+    // 1. The sample can be dropped.
+    // 2. The sample can be recreated and interned into the new profile.
+    // 3. The profile rotation should wait until the sampling operation is complete.
+    //
+    // This API provides a mechanism for samples to pause rotation until they complete, and
+    // for samples to be notified that a rotation is in progress so they can wait to begin.
+    // There are probably better ways, and maybe we should have a notification mechanism.
+    // But for now this should be enough.
+    const FLAG: u64 = u32::MAX as u64;
+
+    /// Prevent any new samples from starting.
+    /// Returns the number of remaining samples.
+    pub fn sample_block(&mut self) -> anyhow::Result<u64> {
+        let current = self.active_samples.fetch_add(Self::FLAG, SeqCst);
+        if current >= Self::FLAG {
+            self.active_samples.fetch_sub(Self::FLAG, SeqCst);
+        }
+        Ok(current % Self::FLAG)
+    }
```

Comment:
Could you explain why this is necessary? 
Also, what would happen when fetch_add overflows?